optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ListAssignMove.java

```diff
@@ -0,0 +1,110 @@
+    private final Object planningValue;
+    private final Object destinationEntity;
+    private final int destinationIndex;
+
+    public ListAssignMove(
+            ListVariableDescriptor<Solution_> variableDescriptor,
+            Object planningValue,
+            Object destinationEntity, int destinationIndex) {
+        this.variableDescriptor = variableDescriptor;
+        this.planningValue = planningValue;
+        this.destinationEntity = destinationEntity;
+        this.destinationIndex = destinationIndex;
+    }
+
+    @Override
+    protected AbstractMove<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        return new ListUnassignMove<>(variableDescriptor, destinationEntity, destinationIndex);
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        // Add element to the destination entity.
+        innerScoreDirector.beforeVariableChanged(variableDescriptor, destinationEntity);
+        variableDescriptor.addElement(destinationEntity, destinationIndex, planningValue);
+        innerScoreDirector.afterVariableChanged(variableDescriptor, destinationEntity);
+    }
+
+    @Override
+    public boolean isMoveDoable(ScoreDirector<Solution_> scoreDirector) {
+        return true;
+    }
+
+    // ************************************************************************
+    // Introspection methods
+    // ************************************************************************
+
+    @Override
+    public String getSimpleMoveTypeDescription() {
+        return getClass().getSimpleName() + "(" + variableDescriptor.getSimpleEntityAndVariableName() + ")";
+    }
+
+    // ************************************************************************
+    // Testing methods
+    // ************************************************************************
+
+    public Object getDestinationEntity() {
+        return destinationEntity;
+    }
+
+    public int getDestinationIndex() {
+        return destinationIndex;
+    }
+
+    public Object getMovedValue() {
+        return planningValue;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof ListAssignMove)) {
+            return false;
+        }
+        ListAssignMove<?> that = (ListAssignMove<?>) o;
+        return destinationIndex == that.destinationIndex && variableDescriptor.equals(that.variableDescriptor)
+                && planningValue.equals(that.planningValue) && destinationEntity.equals(that.destinationEntity);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(variableDescriptor, planningValue, destinationEntity, destinationIndex);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s {unassigned->%s[%d]}",
```

Comment:
1) We want to carefully design these toStrings to be consistent with the other build-in move toStrings,
because they escape into the logs and are a bit therefore something that users get used to being consistent (semi API).
That being said, this looks good, except for a space:

```suggestion
        return String.format("%s {unassigned -> %s[%d]}",
```

2) I like the use of format to make it easier to read. We probably want to use this for older moves too.
OTH, is there a format parsing cost to be paid? So maybe we shouldn't use it here?

Reply:
Actually, in ChangeMove logs, unassigned is written as null IIRC: "Mondayshift { null -> Ann }".
Let's follow that? It's less likely that a (dummy?) shift.toString returns "null" than it does "unassigned", which would lead to confusion.

```suggestion
        return String.format("%s {null -> %s[%d]}",
```

Reply:
> I like the use of format to make it easier to read. We probably want to use this for older moves too.
OTH, is there a format parsing cost to be paid? So maybe we shouldn't use it here?

I don't think the performance of moves' `toString()` methods is an issue. AFAIK, moves are not printed at all on `INFO` level, where you care about the performance.

On `DEBUG` a move is printed once every step, which seems like a good compromise.

Only on `TRACE` level, every move is printed, but that level shouldn't be used in performance-sensitive environments.

Reply:
Proposed format changes incorporated.